in_toto/models/metadata.py

```diff
@@ -198,6 +278,71 @@ def sign_gpg(self, gpg_keyid=None, gpg_home=None):
     return signature
 
 
+  def verify_sigs(self, keys: List[Key], threshold):
+    """Verifies signature over signable in signatures with a list of keys.
+
+    Uses the UTF-8 encoded canonical JSON byte representation of the signable
+    attribute to verify the signature deterministically.
+
+    Arguments:
+        keys: A list of public keys to verify the signatures.
+        threshold: Number of signatures needed to pass the verification.
+
+    Raises:
+        ValueError: If "threshold" is not valid.
+        SignatureVerificationError: If the enclosed signatures do not pass
+            the verification.
+
+    Returns:
+        accepted_keys: A dict of unique public keys.
+    """
+
+    accepted_keys = {}
+    pae = self.signed.signable_bytes
+
+    # checks for threshold value.
+    if threshold <= 0:
+      raise ValueError("Threshold must be greater than 0")
+
+    if len(keys) < threshold:
+      raise ValueError("Number of keys can't be less than threshold")
+
+    for signature in self.signatures:
+
+      # GPGSignature/ Signature from_dict method destroys the signature dict,
+      # Hence, deepcopy of signature is created.
+      if securesystemslib.formats.GPG_SIGNATURE_SCHEMA.matches(signature):
+        signature = GPGSignature.from_dict(copy.deepcopy(signature))
+
+      elif securesystemslib.formats.SIGNATURE_SCHEMA.matches(signature):
+        signature = Signature.from_dict(copy.deepcopy(signature))
+
+      else:
+        raise securesystemslib.exceptions.SignatureVerificationError
```

Comment:
We need to be careful here. A single malformed or unknown signature should not fail the entire verification. We only fail if good signatures don't meet the threshold.

Reply:
This test case was failing that's why I raised this exception.

https://github.com/in-toto/in-toto/blob/8fac55f5bbcfcf36d9c7ff53dddc6c6b12e483ec/tests/test_verifylib.py#L948-L949

Reply:
What did you have there before? IMO that particular test should also fail if you `continue` here instead of raising an error. Then the error is raised further below, because the threshold isn't met. Right?

But it looks like the old `verifylib.verify_layout_signatures` would have also failed for a malformed signature. Interestingly, in both cases it depends on where in the signature list the malformed signature is located, and whether enough valid signatures can be matched by keyid and successfully verified before. It also looks like old and new functions would fail for slightly different sequences of keyid-matching, valid, and malformed signatures .

Regardless, it shouldn't matter for `in_toto_verify`, because `Metablock.validate()` is called on load and will error on malformed signatures before we get to verify them.